### PR TITLE
[dev] add support for deepep/hybridep dispatcher under thd format training

### PR DESCRIPTION
**Pull request by HaochenYuan**
# What does this PR do ?
Previous [PR](https://github.com/NVIDIA/Megatron-LM/pull/2924) added support for thd format in training, but for MoE dispatcher, only all2all type is supported.  This PR adds the deepep & hybridep backend.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/transformer/transformer_config.py [hunk 1]
```diff
@@ -2845,11 +2845,14 @@ def _scope_to_str(s):
...
 
-            # TODO(tailaim): add support for other dispatcher types
-            assert self.moe_token_dispatcher_type == "alltoall", (
-                f"sequence_packing only supports moe_token_dispatcher_type='alltoall', "
-                f"got '{self.moe_token_dispatcher_type}'"
-            )
+            # Dense configs may keep the default MoE dispatcher without ever
+            # constructing MoE layers. For MoE sequence packing, alltoall and
+            # flex dispatchers both support THD-format inputs.
+            if self.num_moe_experts is not None:
```
Comment (kunlunl) [inline megatron/core/transformer/transformer_config.py]: This `if` is not needed?

File: megatron/core/transformer/moe/token_dispatcher.py [hunk 2]
```diff
@@ -1049,24 +1049,45 @@ def __init__(
...
+        self._original_num_tokens: Optional[int] = None
+        self._padded_num_tokens: Optional[int] = None
 
     def setup_metadata(self, routing_map: torch.Tensor, probs: torch.Tensor):
         num_tokens = routing_map.shape[0]
-        self.routing_map = routing_map.reshape(num_tokens, self.num_experts)
-        self.token_probs = probs.reshape(num_tokens, self.num_experts)
+        nt = torch.tensor([num_tokens], device=routing_map.device, dtype=torch.long)
+        torch.distributed.all_reduce(nt, op=torch.distributed.ReduceOp.MAX, group=self.group)
+        padded_num_tokens = int(nt.item())
```
Comment (kunlunl) [inline megatron/core/transformer/moe/token_dispatcher.py]: BSHD doesn't need this line?

**Review by kunlunl — CHANGES_REQUESTED** [review]

**Comment by Victarry** [comment]
@HaochenYuan Could please you add an UT to guard this and submit a PR to main.

cc @Autumn1998 for review

File: megatron/core/transformer/moe/token_dispatcher.py (hunk 2 above)
Comment (HaochenYuan) [reply megatron/core/transformer/moe/token_dispatcher.py]: fixed

File: megatron/core/transformer/transformer_config.py (hunk 1 above)
Comment (HaochenYuan) [reply megatron/core/transformer/transformer_config.py]: fixed

**Comment by HaochenYuan** [comment]
> @HaochenYuan Could please you add an UT to guard this and submit a PR to main.
> 
> cc @Autumn1998 for review

UT added.

**Review by kunlunl — APPROVED** [review]

File: tests/unit_tests/transformer/moe/test_token_dispatcher.py [hunk 3]
```diff
@@ -461,6 +468,181 @@ def is_hybrid_ep_available():
...
+        tp_rank * sequence_parallel_length : (tp_rank + 1) * sequence_parallel_length
+    ]
+    return hidden_states.unsqueeze(1).contiguous().requires_grad_(True)
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+@pytest.mark.skipif(
+    Utils.world_size % 8 != 0,
+    reason="requires world size divisible by 8 for pp2/cp2/tp2/ep2/etp2",
+)
```
[the remark this answers is not in the document]
Comment (HaochenYuan) [reply tests/unit_tests/transformer/moe/test_token_dispatcher.py:547]: cp won't affect the moe part, the needed gpu count is 8.

File: megatron/core/transformer/moe/token_dispatcher.py [hunk 4]
```diff
@@ -1049,24 +1049,50 @@ def __init__(
...
-        self.token_probs = probs.reshape(num_tokens, self.num_experts)
+        self._original_num_tokens = num_tokens
+
+        padded_num_tokens = num_tokens
+        if self.config.sequence_packing_scheduler is not None:
+            # Use the actual tp_ep max so all ranks in the MoE communication
+            # group pass the same token count to HybridEP.
+            nt = torch.tensor([num_tokens], device=routing_map.device, dtype=torch.long)
+            torch.distributed.all_reduce(nt, op=torch.distributed.ReduceOp.MAX, group=self.group)
+            padded_num_tokens = int(nt.item())
```
[the remark this answers is not in the document]
Comment (HaochenYuan) [reply megatron/core/transformer/moe/token_dispatcher.py]: The code here doesn't consider to be graph-compatible, we will add cuda graph support elsewhere.

**Review by yaox12 — APPROVED** [review]
LGTM. Please fix the CI failure.

File: megatron/core/transformer/moe/token_dispatcher.py [hunk 5]
```diff
@@ -1049,24 +1049,51 @@ def __init__(
...
+            _HYBRIDEP_TOKEN_ALIGNMENT = 64
+            padded_num_tokens += -padded_num_tokens % _HYBRIDEP_TOKEN_ALIGNMENT
```
Comment (Victarry) [inline megatron/core/transformer/moe/token_dispatcher.py]: 1. Please don't use very short variable name like `nt`. Something like `max_num_token_across_ep` is more informative
2. _HYBRIDEP_TOKEN_ALIGNMENT is hardcoded, it should be placed in hybridep related code in fused_a2a.py

**Review by Autumn1998 — APPROVED** [review]
LGTM

File: megatron/core/transformer/moe/token_dispatcher.py (hunk 5 above)
Comment (HaochenYuan) [reply megatron/core/transformer/moe/token_dispatcher.py]: solved

**Review by Victarry — APPROVED** [review]
